zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -219,44 +297,169 @@ where
+            recipient_address: &payment.recipient_address,
+            value: payment.amount,
+            memo: payment.memo.clone()
         }
-    };
+    }).collect();
 
     wallet_db.store_sent_tx(&SentTransaction {
         tx: &tx,
         created: time::OffsetDateTime::now_utc(),
-        output_index,
+        outputs: sent_outputs,
         account,
-        recipient_address: to,
-        value,
-        memo,
+        utxos_spent: vec![],
+    })
+}
+
+/// Constructs a transaction that consumes available transparent UTXOs belonging to
+/// the specified secret key, and sends them to the default address for the provided Sapling
+/// extended full viewing key.
+///
+/// This procedure will not attempt to shield transparent funds if the total amount being shielded
+/// is less than the default fee to send the transaction. Fees will be paid only from the transparent
+/// UTXOs being consumed.
+///
+/// Parameters:
+/// * `wallet_db`: A read/write reference to the wallet database
+/// * `params`: Consensus parameters
+/// * `prover`: The TxProver to use in constructing the shielded transaction.
+/// * `sk`: The secp256k1 secret key that will be used to detect and spend transparent
+///   UTXOs.
+/// * `extfvk`: The extended full viewing key that will be used to produce the
+///   Sapling address to which funds will be sent.
+/// * `account`: The ZIP32 account identifier associated with the the extended
+///   full viewing key. This procedure will return an error if this does not correctly
+///   correspond to `extfvk`.
+/// * `min_confirmations`: The minimum number of confirmations that a previously
+///   received UTXO must have in the blockchain in order to be considered for being
+///   spent.
+#[cfg(feature = "transparent-inputs")]
+#[allow(clippy::too_many_arguments)]
+pub fn shield_transparent_funds<E, N, P, D, R>(
+    wallet_db: &mut D,
+    params: &P,
+    prover: impl TxProver,
+    sk: &secp256k1::SecretKey,
+    extfvk: &ExtendedFullViewingKey,
+    account: AccountId,
+    memo: &MemoBytes,
+    min_confirmations: u32,
+) -> Result<D::TxRef, E>
+where
+    E: From<Error<N>>,
+    P: consensus::Parameters,
+    R: Copy + Debug,
+    D: WalletWrite<Error = E, TxRef = R>,
+{
+    // Check that the ExtendedSpendingKey we have been given corresponds to the
+    // ExtendedFullViewingKey for the account we are spending from.
+    if !wallet_db.is_valid_account_extfvk(account, &extfvk)? {
+        return Err(E::from(Error::InvalidExtSk(account)));
+    }
+
+    let (latest_scanned_height, latest_anchor) = wallet_db
+        .get_target_and_anchor_heights(min_confirmations)
+        .and_then(|x| x.ok_or_else(|| Error::ScanRequired.into()))?;
+
+    // derive the corresponding t-address
+    let taddr = derive_transparent_address_from_secret_key(sk);
+
+    // derive own shielded address from the provided extended spending key
+    let z_address = extfvk.default_address().1;
+    let ovk = extfvk.fvk.ovk;
```

Comment:
Why are we using the ovk from the destination?

Reply:
We won't need to now that we have ZIP-316 specifying how to derive OVKs for transparent addresses.

Reply:
OK. Fortunately this doesn't appear to have been a severe security issue for existing transfers. It effectively grants incoming viewing authority to the holder of an ovk key for the destination address. But since we don't have a user-visible abstraction for separate OVKs, the ovk key is normally bundled only with the FVK for the destination address, which is intended to grant incoming viewing authority.